File: backend/measure.py
"""Measurement and quantity take-off.

What separates these from ordinary shapes is the scale: every measurement is a
drawing distance multiplied by a ratio the user calibrates once. The result is
stored on the annotation so it can be recomputed and summed later.
"""

from __future__ import annotations

import math

# Conversion to millimetres. Points are the PDF's own unit (72 per inch).
UNITS = {
    "pt": 25.4 / 72,
    "mm": 1.0,
    "cm": 10.0,
    "m": 1000.0,
    "in": 25.4,
    "ft": 304.8,
}

AREA_UNITS = {"mm": "mm²", "cm": "cm²", "m": "m²", "in": "in²", "ft": "ft²", "pt": "pt²"}


def scale_factor(scale: dict) -> float:
    """Drawing points to real-world units of `scale['unit']`.

    A scale says "this many points on the page equals this much in reality",
    which is what calibrating against a known dimension produces.
    """
    page_length = float(scale.get("pagePoints") or 1)
    real_length = float(scale.get("realLength") or 1)
    if page_length <= 0:
        return 1.0
    return real_length / page_length


def calibrate(p1, p2, real_length: float, unit: str) -> dict:
    """Build a scale from a line the user drew over a known dimension."""
    points = math.dist(p1, p2)
    return {
        "pagePoints": points,
        "realLength": real_length,
        "unit": unit if unit in UNITS else "mm",
    }


def distance(points: list[list[float]], scale: dict) -> float:
    total = 0.0
    for a, b in zip(points, points[1:]):
        total += math.dist(a, b)
    return total * scale_factor(scale)


def perimeter(points: list[list[float]], scale: dict, closed: bool = True) -> float:
    loop = points + [points[0]] if closed and len(points) > 2 else points
    return distance(loop, scale)


def area(points: list[list[float]], scale: dict) -> float:
    """Shoelace area, converted to square units of the scale."""
    if len(points) < 3:
        return 0.0
    total = 0.0
    for (x0, y0), (x1, y1) in zip(points, points[1:] + [points[0]]):
        total += x0 * y1 - x1 * y0
    factor = scale_factor(scale)
    return abs(total) / 2 * factor * factor


def angle(points: list[list[float]]) -> float:
    """Angle at the middle point of three, in degrees."""
    if len(points) < 3:
        return 0.0
    (ax, ay), (bx, by), (cx, cy) = points[:3]
    first = math.atan2(ay - by, ax - bx)
    second = math.atan2(cy - by, cx - bx)
    degrees = math.degrees(abs(first - second))
    return 360 - degrees if degrees > 180 else degrees


def radius(points: list[list[float]], scale: dict) -> float:
    if len(points) < 2:
        return 0.0
    return math.dist(points[0], points[1]) * scale_factor(scale)


def volume(points: list[list[float]], scale: dict, depth: float) -> float:
    return area(points, scale) * depth


def format_value(value: float, kind: str, scale: dict, precision: int = 2) -> str:
    unit = scale.get("unit", "mm")
    if kind == "angle":
        return f"{value:.{precision}f}°"
    if kind == "area":
        return f"{value:.{precision}f} {AREA_UNITS.get(unit, unit + '²')}"
    if kind == "volume":
        return f"{value:.{precision}f} {unit}³"
    return f"{value:.{precision}f} {unit}"


def measure(kind: str, points: list[list[float]], scale: dict, *,
            depth: float = 0.0, precision: int = 2) -> dict:
    """Compute one measurement and return both the number and its label."""
    if kind == "distance":
        value = distance(points, scale)
    elif kind == "perimeter":
        value = perimeter(points, scale)
    elif kind == "area":
        value = area(points, scale)
    elif kind == "angle":
        value = angle(points)
    elif kind == "radius":
        value = radius(points, scale)
    elif kind == "volume":
        value = volume(points, scale, depth)
    else:
        raise ValueError(f"未対応の計測種別: {kind}")
    return {
        "kind": kind,
        "value": round(value, 6),
        "unit": scale.get("unit", "mm"),
        "label": format_value(value, kind, scale, precision),
    }


def convert(value: float, from_unit: str, to_unit: str) -> float:
    return value * UNITS.get(from_unit, 1.0) / UNITS.get(to_unit, 1.0)


# ---------------------------------------------------------------- take-off


KIND_LABELS = {
    "distance": "距離", "perimeter": "周囲長", "area": "面積",
    "angle": "角度", "radius": "半径", "volume": "体積", "count": "カウント",
}


def _unit_for(kind: str, unit: str) -> str:
    if kind == "count":
        return ""
    if kind == "angle":
        return "°"
    if kind == "area":
        return AREA_UNITS.get(unit, f"{unit}²")
    if kind == "volume":
        return f"{unit}³"
    return unit


def summarise(items: list[dict]) -> dict:
    """Group measurements and counts into a take-off table.

    Grouping is by subject, which is the field reviewers already use to label
    what a mark refers to, so the legend and the totals stay in step.
    """
    groups: dict[str, dict] = {}
    for item in items:
        measurement = item.get("measure")
        is_count = item.get("tool") == "count"
        if not measurement and not is_count:
            continue
        kind = "count" if is_count else measurement.get("kind", "")
        key = item.get("subject") or KIND_LABELS.get(kind, kind)
        group = groups.setdefault(key, {
            "label": key,
            "colour": (item.get("style") or {}).get("stroke", "#888888"),
            "kind": kind,
            "unit": _unit_for(kind, (measurement or {}).get("unit", "")),
            # Adding up angles is meaningless; only their count is reported.
            "summable": kind not in ("count", "angle"),
            "count": 0,
            "total": 0.0,
            "pages": set(),
        })
        group["count"] += 1
        group["pages"].add(int(item.get("page", 0)) + 1)
        if measurement:
            group["total"] += float(measurement.get("value") or 0)

    rows = []
    for group in groups.values():
        rows.append({
            **group,
            "total": round(group["total"], 4),
            "pages": sorted(group["pages"]),
        })
    rows.sort(key=lambda r: r["label"])
    return {"rows": rows, "groups": len(rows), "items": sum(r["count"] for r in rows)}


def to_csv(summary: dict, items: list[dict]) -> str:
    import csv
    import io

    buffer = io.StringIO()
    writer = csv.writer(buffer, lineterminator="\n")
    writer.writerow(["分類", "種別", "個数", "合計", "単位", "ページ", "色"])
    for row in summary["rows"]:
        writer.writerow([
            row["label"], row["kind"], row["count"], row["total"], row["unit"],
            " ".join(str(p) for p in row["pages"]), row["colour"],
        ])
    writer.writerow([])
    writer.writerow(["ページ", "分類", "種別", "値", "単位", "表示", "コメント"])
    for item in sorted(items, key=lambda i: (i.get("page", 0), i.get("rect", [0, 0])[1])):
        measurement = item.get("measure")
        if not measurement and item.get("tool") != "count":
            continue
        writer.writerow([
            int(item.get("page", 0)) + 1,
            item.get("subject", ""),
            (measurement or {}).get("kind", "count"),
            (measurement or {}).get("value", 1),
            (measurement or {}).get("unit", ""),
            (measurement or {}).get("label", ""),
            item.get("contents", ""),
        ])
    return buffer.getvalue()

File: webapp/js/measure.js
// Measuring and counting, in the browser.
//
// The same maths runs on the server for take-off totals; here it runs live so
// the value follows the cursor while the user is still drawing.

const UNITS = { pt: 25.4 / 72, mm: 1, cm: 10, m: 1000, in: 25.4, ft: 304.8 };
const AREA_UNITS = { mm: 'mm²', cm: 'cm²', m: 'm²', in: 'in²', ft: 'ft²', pt: 'pt²' };

const KEY = 'pdfstudio.scale.v1';

const DEFAULT_SCALE = { pagePoints: 1, realLength: 1, unit: 'mm' };

let scale;
try {
  scale = { ...DEFAULT_SCALE, ...JSON.parse(localStorage.getItem(KEY) || '{}') };
} catch { scale = { ...DEFAULT_SCALE }; }

export function getScale() {
  return { ...scale };
}

export function setScale(next) {
  scale = { ...scale, ...next };
  try { localStorage.setItem(KEY, JSON.stringify(scale)); } catch { /* private mode */ }
}

/** Is the scale still the untouched 1pt = 1mm placeholder? */
export function isCalibrated() {
  return !(scale.pagePoints === 1 && scale.realLength === 1);
}

export function factor() {
  return scale.pagePoints > 0 ? scale.realLength / scale.pagePoints : 1;
}

export function calibrateFrom(p1, p2, realLength, unit) {
  const points = Math.hypot(p2[0] - p1[0], p2[1] - p1[1]);
  setScale({ pagePoints: points, realLength, unit });
  return getScale();
}

function pathLength(points) {
  let total = 0;
  for (let i = 1; i < points.length; i += 1) {
    total += Math.hypot(points[i][0] - points[i - 1][0], points[i][1] - points[i - 1][1]);
  }
  return total;
}

function polygonArea(points) {
  if (points.length < 3) return 0;
  let total = 0;
  for (let i = 0; i < points.length; i += 1) {
    const [x0, y0] = points[i];
    const [x1, y1] = points[(i + 1) % points.length];
    total += x0 * y1 - x1 * y0;
  }
  return Math.abs(total) / 2;
}

function angleAt(points) {
  if (points.length < 3) return 0;
  const [[ax, ay], [bx, by], [cx, cy]] = points;
  const first = Math.atan2(ay - by, ax - bx);
  const second = Math.atan2(cy - by, cx - bx);
  const degrees = Math.abs(first - second) * 180 / Math.PI;
  return degrees > 180 ? 360 - degrees : degrees;
}

export function compute(kind, points, { depth = 0, precision = 2 } = {}) {
  const f = factor();
  let value;
  if (kind === 'distance') value = pathLength(points) * f;
  else if (kind === 'perimeter') value = pathLength([...points, points[0]]) * f;
  else if (kind === 'area') value = polygonArea(points) * f * f;
  else if (kind === 'volume') value = polygonArea(points) * f * f * depth;
  else if (kind === 'angle') value = angleAt(points);
  else if (kind === 'radius') value = pathLength(points.slice(0, 2)) * f;
  else value = 0;
  return {
    kind,
    value: Number(value.toFixed(6)),
    unit: scale.unit,
    label: format(value, kind, precision),
  };
}

export function format(value, kind, precision = 2) {
  if (kind === 'angle') return `${value.toFixed(precision)}°`;
  if (kind === 'area') return `${value.toFixed(precision)} ${AREA_UNITS[scale.unit] || `${scale.unit}²`}`;
  if (kind === 'volume') return `${value.toFixed(precision)} ${scale.unit}³`;
  return `${value.toFixed(precision)} ${scale.unit}`;
}

export function convert(value, from, to) {
  return value * (UNITS[from] || 1) / (UNITS[to] || 1);
}

export const UNIT_LABELS = {
  mm: 'ミリメートル (mm)',
  cm: 'センチメートル (cm)',
  m: 'メートル (m)',
  in: 'インチ (in)',
  ft: 'フィート (ft)',
  pt: 'ポイント (pt)',
};

export const KIND_LABELS = {
  distance: '距離', perimeter: '周囲長', area: '面積',
  angle: '角度', radius: '半径', volume: '体積', count: 'カウント',
};

/** The unit a total should carry — angles are degrees, counts have none. */
function unitFor(kind, unit) {
  if (kind === 'count') return '';
  if (kind === 'angle') return '°';
  if (kind === 'area') return AREA_UNITS[unit] || `${unit}²`;
  if (kind === 'volume') return `${unit}³`;
  return unit;
}

/** Group measurements and counts the way the server does, for the live panel. */
export function summarise(items) {
  const groups = new Map();
  for (const item of items) {
    const isCount = item.tool === 'count';
    if (!item.measure && !isCount) continue;
    const kind = isCount ? 'count' : item.measure.kind;
    const key = item.subject || KIND_LABELS[kind] || kind;
    if (!groups.has(key)) {
      groups.set(key, {
        label: key,
        colour: item.style?.stroke || '#888888',
        kind,
        unit: unitFor(kind, item.measure?.unit),
        // Adding up angles is meaningless; only the count is shown for them.
        summable: kind !== 'count' && kind !== 'angle',
        count: 0,
        total: 0,
        pages: new Set(),
      });
    }
    const group = groups.get(key);
    group.count += 1;
    group.pages.add(item.page + 1);
    if (item.measure) group.total += Number(item.measure.value) || 0;
  }
  return [...groups.values()]
    .map((g) => ({ ...g, total: Number(g.total.toFixed(4)), pages: [...g.pages].sort((a, b) => a - b) }))
    .sort((a, b) => a.label.localeCompare(b.label, 'ja'));
}

export const MEASURE_KINDS = {
  measureDistance: 'distance',
  measurePerimeter: 'perimeter',
  measureArea: 'area',
  measureAngle: 'angle',
  measureRadius: 'radius',
};
